lib/propolis/src/enlightenment/hyperv/mod.rs

```diff
@@ -0,0 +1,313 @@
+                self.inner.lock().unwrap().msr_hypercall_value.0,
+            ),
+            HV_X64_MSR_VP_INDEX => {
+                let id: u32 = vcpu.into();
+                RdmsrOutcome::Handled(id as u64)
+            }
+            _ => RdmsrOutcome::NotHandled,
+        }
+    }
+
+    fn wrmsr(&self, _vcpu: VcpuId, msr: MsrId, value: u64) -> WrmsrOutcome {
+        match msr.0 {
+            HV_X64_MSR_GUEST_OS_ID => self.handle_wrmsr_guest_os_id(value),
+            HV_X64_MSR_HYPERCALL => self.handle_wrmsr_hypercall(value),
+            HV_X64_MSR_VP_INDEX => WrmsrOutcome::GpException,
+            _ => WrmsrOutcome::NotHandled,
+        }
+    }
+
+    fn attach(&self, mem_acc: &MemAccessor) {
+        mem_acc.adopt(&self.acc_mem, Some(TYPE_NAME.to_owned()));
+    }
+}
+
+fn write_overlay_page(mapping: &SubMapping<'_>, contents: &[u8; PAGE_SIZE]) {
+    let written = mapping
+        .write_bytes(contents)
+        .expect("overlay pages are always writable");
+
+    assert_eq!(written, PAGE_SIZE, "overlay pages can be written completely");
+}
+
+impl Lifecycle for HyperV {
+    fn type_name(&self) -> &'static str {
+        TYPE_NAME
+    }
+
+    fn reset(&self) {
+        let mut inner = self.inner.lock().unwrap();
+        *inner = Inner::default();
+    }
+
+    fn migrate(&'_ self) -> Migrator<'_> {
+        Migrator::Single(self)
+    }
+}
+
+impl MigrateSingle for HyperV {
+    fn export(
+        &self,
+        _ctx: &MigrateCtx,
+    ) -> Result<PayloadOutput, MigrateStateError> {
+        let inner = self.inner.lock().unwrap();
+        Ok(migrate::HyperVEnlightenmentV1 {
+            msr_guest_os_id: inner.msr_guest_os_id_value,
+            msr_hypercall: inner.msr_hypercall_value.0,
+        }
+        .into())
+    }
+
+    fn import(
+        &self,
+        mut offer: PayloadOffer,
+        ctx: &MigrateCtx,
+    ) -> Result<(), MigrateStateError> {
+        let data: migrate::HyperVEnlightenmentV1 = offer.parse()?;
+
+        // A well-behaved source should ensure that the hypercall MSR value is
+        // within the guest's PA range and that its Enabled bit agrees with the
+        // value of the guest OS ID MSR. But this data was received over the
+        // wire, so for safety's sake, verify it all and return a migration
+        // error if anything is inconsistent.
+        let hypercall_msr = MsrHypercallValue(data.msr_hypercall);
+        if hypercall_msr.enabled() {
+            if data.msr_guest_os_id == 0 {
+                return Err(MigrateStateError::ImportFailed(
+                    "hypercall MSR enabled but guest OS ID MSR is 0"
+                        .to_string(),
+                ));
+            }
+
+            let Some(mapping) = ctx
+                .mem
+                .writable_region(&GuestRegion(hypercall_msr.gpa(), PAGE_SIZE))
+            else {
+                return Err(MigrateStateError::ImportFailed(
+                    "couldn't map GPA in hypercall MSR".to_string(),
```

Comment:
nit, take it or leave it: is it worth including the specified GPA here? we're `to_string`ing anyway so we may as well `format!`...

Reply:
Fixed in 5bf277f7.